### PR TITLE
BLE upgrades

### PR DESCRIPTION
**Pull request by PilnyTomas**
### General:
**BREAKING CHANGE!** Changed API return and parameter type from `std::tring` to Arduino style `String`.
Renamed library name from `ESP32 BLE Arduino` to simple `BLE`.
Created a new function `getFrame` which returns complete data for `BLEAdvertisementData::addData()`.
Renamed example files - removed prefix `BLE_` - `BLE5_` is left unchanged.
`Beacon_Scanner` example updated to reflect changes in API - shortened by 28 lines on TLM and URL part.
Changed UUID data type from `uint16_t` to `BLEUUID` class

### Eddysstone URL:
Created a new constructor which takes as an argument a `BLEAdvertisedDevice` object and initializes URL data from the payload.
Created a new function `BLEEddystoneURL::setSmartURL` which encodes the full URL structure.

Static frame data no longer needs to be written by the user - they are initialized by the constructor.
Function `BLEEddystoneURL::setPower` now accepts `esp_power_level_t`.
Updated `EddystoneURL_Beacon` example to reflect changes in API - shortened by **83** lines.
Added comments with explanations to functions `setData` and `setURL`.

**Tests:**
`EddystoneURL_Beacon` with the following URLs and read them in `Beacon_Scanner` example and using an Android app [nRF Connect for Mobile](https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp&hl=en&gl).
```
char unprintable[] = {0x01, 0xFF, 0xDE, 0xAD};
String URL[] = {
"http://www.espressif.com/", // prefix 0x00, suffix 0x00
"https://www.texas.gov",     // prefix 0x01, suffix 0x0D
"http://en.mapy.cz",         // prefix 0x02, no valid suffix
"https://arduino.cc",        // prefix 0x03, no valid suffix
"google.com",                // URL without specified prefix - the function will assume default prefix "http://www." = 0x00
"diginfo.tv",                // URL without specified prefix - the function will assume default prefix "http://www." = 0x00
"http://www.URLsAbove17BytesAreNotAllowed.com", // Too long URL - setSmartURL() will return 0 = ERR
String(unprintable) // Unprintable characters - setSmartURL() will return 0 = ERR
```

### Eddystone TLM:
Created a new constructor which takes as an argument a `BLEAdvertisedDevice` object and initializes TLM data from the payload.
Added comments with explanations to functions `setData` and `setVolt`

TODO: test all examples if they are properly working - changes related to the String might have broken something!

Related task: #7841

### CODE REVIEW

**Comment by SuGlider** [comment]
@PilnyTomas - There are commits from 2022 in this PR. Is it right?

image

**Comment by PilnyTomas** [comment]
> There are commits from 2022 in this PR. Is it right?

To b honest, I don't know how they got there, but considering that they are mostly just merges of master it shouldn't be a problem. The only change was adding the vector #include, but as is not needed I have removed it in one of the following commits.

File: libraries/BLE/src/RTOS.h [hunk 1]
```diff
@@ -24,31 +24,31 @@
...
-		bool        take(std::string owner = "<Unknown>");
-		bool        take(uint32_t timeoutMs, std::string owner = "<Unknown>");
-		std::string toString();
-		uint32_t	wait(std::string owner = "<Unknown>");
-		bool		timedWait(std::string owner = "<Unknown>", uint32_t timeoutMs = portMAX_DELAY);
+		void        setName(String name);
+		bool        take(String owner = "<Unknown>");
+		bool        take(uint32_t timeoutMs, String owner = "<Unknown>");
+		String toString();
+		uint32_t	wait(String owner = "<Unknown>");
```
Comment (lucasssvaz) [inline libraries/BLE/src/RTOS.h:43]:
```suggestion
		bool        take(uint32_t timeoutMs, String owner = "<Unknown>");
		String      toString();
		uint32_t	wait(String owner = "<Unknown>");
```

File: libraries/BLE/src/RTOS.h [hunk 2]
```diff
@@ -24,31 +24,31 @@
...
+		bool		timedWait(String owner = "<Unknown>", uint32_t timeoutMs = portMAX_DELAY);
 		uint32_t	value(){ return m_value; };
 
 	private:
 		SemaphoreHandle_t m_semaphore;
 		pthread_mutex_t   m_pthread_mutex;
-		std::string       m_name;
-		std::string       m_owner;
+		String       m_name;
+		String       m_owner;
```
Comment (lucasssvaz) [inline libraries/BLE/src/RTOS.h:51]:
```suggestion
		String            m_name;
		String            m_owner;
```

File: libraries/BLE/src/BLEAdvertisedDevice.h [hunk 3]
```diff
@@ -9,17 +9,25 @@
...
 #include "soc/soc_caps.h"
 #if SOC_BLE_SUPPORTED
+
 #include "sdkconfig.h"
 #if defined(CONFIG_BLUEDROID_ENABLED)
 #include <esp_gattc_api.h>
 
 #include <map>
-#include <vector>
+//#include <vector>
```
Comment (lucasssvaz) [inline libraries/BLE/src/BLEAdvertisedDevice.h:18]:
```suggestion

```

**Comment by VojtechBartoska** [comment]
Related to the BLE refactoring: https://github.com/espressif/arduino-esp32/issues/7702

**Comment by me-no-dev** [comment]
Closed in favor of https://github.com/espressif/arduino-esp32/pull/8724

**Pull request closed without merging**